src/ucs/sys/topo/nvml/topo_nvml.c

```diff
@@ -0,0 +1,410 @@
+    }
+
+    bw = (bus_width / 8) * bus_clock * 1024.0 * 1024.0;
+
+    ucs_trace("device bus width %u bus clock %u bw %lf", bus_width, bus_clock, bw);
+
+    return bw;
+
+exit:
+    bw = 0.0;
+    return bw;
+}
+
+/* report peak host<->device bandwidth */
+static double ucs_sys_nvml_get_pci_bw(nvmlDevice_t *device)
+{
+    double bw = 0.0; /* set default bw here */
+    unsigned link_gen;
+    unsigned link_width;
+    nvmlReturn_t nvml_err;
+
+
+    nvml_err = nvmlDeviceGetCurrPcieLinkGeneration(*device, &link_gen);
+    if (nvml_err != NVML_SUCCESS) {
+        goto exit;
+    }
+
+    nvml_err = nvmlDeviceGetCurrPcieLinkWidth(*device, &link_width);
+    if (nvml_err != NVML_SUCCESS) {
+        goto exit;
+    }
+
+    ucs_trace("nvml device link gen = %d, link width = %d", link_gen, link_width);
+
+    switch(link_gen) {
+        case 1:
+            bw = link_width * 250 * UCS_MBYTE;
+            break;
+        case 2:
+            bw = link_width * 500 * UCS_MBYTE;
+            break;
+        case 3:
+            bw = link_width * 985 * UCS_MBYTE;
+            break;
+        case 4:
+            bw = link_width * 1970 * UCS_MBYTE;
+            break;
+        case 5:
+            bw = link_width * 3940 * UCS_MBYTE;
+            break;
+        default:
+            bw = 0.0; /* set default bw here */
+            break;
+    }
+
+exit:
+    return bw;
+}
+
+static ucs_status_t
+ucs_topo_get_distance_nvml(ucs_sys_device_t local_sys_device,
+                           ucs_sys_device_t remote_sys_device,
+                           ucs_sys_dev_distance_t *distance)
+{
+    double bw                  = 0.0;
+    double lat                 = 1e-9;
+    nvmlDevice_t local_device  = 0;
+    nvmlDevice_t remote_device = 0;
+    char local_device_name[UCS_SYS_TOPO_NVML_DEV_LEN];
+    char remote_device_name[UCS_SYS_TOPO_NVML_DEV_LEN];
+    ucs_sys_topo_common_ancestor_t common_ancestor;
+    ucs_status_t status;
+
+    strcpy(local_device_name, ucs_topo_sys_device_get_name(local_sys_device));
+    strcpy(remote_device_name, ucs_topo_sys_device_get_name(remote_sys_device));
+
+    ucs_sys_nvml_get_nvml_device(local_sys_device, &local_device);
+    ucs_sys_nvml_get_nvml_device(remote_sys_device, &remote_device);
+
+    if ((!strncmp(local_device_name, "GPU", 3) && (remote_sys_device == UCS_SYS_DEVICE_ID_UNKNOWN)) ||
+        (!strncmp(remote_device_name, "GPU", 3) && (local_sys_device == UCS_SYS_DEVICE_ID_UNKNOWN))) {
+        /* assume pci between local and remote for now */
+        goto pci_bw;
+    } else if (!strncmp(remote_device_name, "GPU", 3) && !strncmp(local_device_name, "GPU", 3)) {
+        if (local_device != remote_device) {
+            status = ucs_sys_nvml_get_common_ancestor(local_device, remote_device,
+                    &common_ancestor);
+            if (status == UCS_OK) {
+                if (common_ancestor != UCS_SYS_TOPO_COMMON_SYSTEM) {
+                    /* both devices in the same numa domain */
```

Comment:
Why is this needed? devices under two different NUMA nodes can still be connected via NVLink.